Add MYNN surface layer scheme

kuanchihwang opened the pull request:
### Originator(s):

kuanchihwang

### Descriptions (include the issue title, and the keyword ['closes', 'fixes', 'resolves'] followed by the issue number):

This PR adds the Mellor-Yamada-Nakanishi-Niino (MYNN) surface layer scheme to the experimental convection-permitting physics suite.

### List all namelist files that were added or changed:

None

### List all files eliminated and why:

None

### List all files added and what they do:

```
A       .gitattributes
  * Make GitHub Linguist recognize `*.pf` files as Fortran source code
A       schemes/mmm/sf_mynn_compat.F90
A       schemes/mmm/sf_mynn_compat.meta
  * Implement MYNN surface layer scheme
```

### List all existing files that have been modified, and describe the changes:

```
M       .github/workflows/unit-tests.yaml
  * Do not hard code path to working directory in GitHub Actions
M       .gitmodules
M       schemes/mmm/mmm_physics
  * Update git submodule of MMM physics
M       schemes/mmm/CMakeLists.txt
M       test/unit-test/tests/mmm/CMakeLists.txt
  * Synchronize `CMakeLists.txt` of MMM physics with MPAS dycore in CAM-SIMA
M       schemes/mmm/cu_ntiedtke_compat.F90
M       schemes/mmm/cu_ntiedtke_compat.meta
  * Fix incorrect standard name in new Tiedtke convection scheme
M       schemes/mmm/mmm_physics_compat.F90
M       schemes/mmm/mmm_physics_compat.meta
  * Implement interstitial schemes for MYNN surface layer scheme
M       test/test_suites/suite_convection_permitting.xml
  * Add MYNN surface layer scheme to convection-permitting suite
M       test/unit-test/tests/mmm/mmm_physics_compat_tests.pf
  * Implement unit tests
```

### List all automated tests that failed, as well as an explanation for why they were not fixed:

None

### Is this an answer-changing PR? If so, is it a new physics package, algorithm change, tuning change, etc?

Answer-changing for the convection-permitting physics suite due to a newly added physics scheme. Nothing is changed for the rest.

### If yes to the above question, describe how this code was validated with the new/modified features:

The convection-permitting physics suite is considered an experimental feature. There is no baseline available to validate against because it has never been implemented in CAM-SIMA as well as CAM before.

kuanchihwang committed:
Do not hard code path to working directory in GitHub Actions

The workflow will fail when a fork uses a different repository name.

kuanchihwang committed:
Fix incorrect standard name in new Tiedtke convection scheme

`hfx` is simply the sensible heat flux, not the sum of latent heat flux
and sensible heat flux.

kuanchihwang committed:
Implement interstitial schemes for MYNN surface layer scheme

kuanchihwang committed:
Implement MYNN surface layer scheme

kuanchihwang committed:
Add MYNN surface layer scheme to convection-permitting suite

kuanchihwang committed:
Implement unit tests

kuanchihwang committed:
Make GitHub Linguist recognize `*.pf` files as Fortran source code

kuanchihwang committed:
Synchronize `CMakeLists.txt` of MMM physics with MPAS dycore in CAM-SIMA

kuanchihwang committed:
Update git submodule of MMM physics

... for fixing compilation errors with MYNN surface layer scheme.

kuanchihwang committed:
Merge branch 'development' into develop/mynn-sf-scheme

nusbaume commented on schemes/mmm/sf_mynn_compat.F90:
```diff
@@ -0,0 +1,513 @@
+            spp_pbl, rstoch1d, &
+            errmsg, errflg)
+        use ccpp_kinds, only: kind_phys
+        use ccpp_scheme_utils, only: ccpp_constituent_index
```
It doesn't look like `ccpp_constituent_index` is used in this subroutine?

nusbaume commented on schemes/mmm/sf_mynn_compat.F90:
```diff
@@ -0,0 +1,513 @@
+            itimestep, wstar, qstar, ustm, ck, cka, &
+            cd, cda, spp_pbl, rstoch1d, isftcflx, &
+            iz0tlnd, its, ite, &
+            errmsg, errflg)
```
Should we check `errflg` here?


```suggestion
            errmsg, errflg)
            
            if (errflg /= 0) then
                return
            end if
```

nusbaume commented on schemes/mmm/sf_mynn_compat.F90:
```diff
@@ -0,0 +1,513 @@
+                itimestep, wstar_sea, qstar_sea, ustm_sea, ck_sea, cka_sea, &
+                cd_sea, cda_sea, spp_pbl, rstoch1d, isftcflx, &
+                iz0tlnd, its, ite, &
+                errmsg, errflg)
```
Should we check `errflg` here as well?


```suggestion
                errmsg, errflg)
                
                if (errflg /= 0) then
                    return
                end if
```

kuanchihwang replied:
I added more checks for error codes in https://github.com/ESCOMP/atmospheric_physics/pull/340/commits/c2959dfcd0197ba7803c95878cba1c6867d5da96.

nusbaume commented on test/test_suites/suite_convection_permitting.xml:
```diff
@@ -13,6 +15,10 @@
 
     <scheme>mmm_physics_compat</scheme>
 
+    <scheme>sf_mynn_compat_pre</scheme>
```
Given that this suite is getting bigger, can we add comments to the set of scheme calls for each parameterization?  For example:


```suggestion
    <!-- Mellor–Yamada–Nakanishi–Niino (MYNN) Surface Layer -->
    <scheme>sf_mynn_compat_pre</scheme>
```

You can also see examples of these comments in the [CAM4 suite](https://github.com/ESCOMP/atmospheric_physics/blob/main/suites/suite_cam4.xml) as well.

nusbaume reviewed (changes requested):
Thanks for bringing in the first MYNN scheme @kuanchihwang!  I do have some change requests, mostly related to the location of certain subroutines, the conversion of hard-coded variables to namelist variables, and a few other general requests.  Of course if you have any questions or concerns with my requests just let me know.  Thanks again!

kuanchihwang committed:
Merge branch 'development' into develop/mynn-sf-scheme

kuanchihwang committed:
Remove unused variable

kuanchihwang committed:
Move parameter to module level to avoid multiple declarations

kuanchihwang committed:
Add more checks for error flags

kuanchihwang committed:
Update standard name to be more clear

However, this standard name change deviates from the ESM Standard Name Library.

In the ESM Standard Name Library, "ratio_of_height_to_monin_obukhov_length" is
defined. However, here "ratio_of_height_at_surface_adjacent_layer_to_monin_obukhov_length"
is used.

kuanchihwang committed:
Add comments to suite definition file

kuanchihwang committed:
Move computational procedures that are general enough to state converters

* To conform to the existing naming convention in state converters,
  `compute_ ...` has been renamed to `calc_ ...`.
* Fix incorrect units in a code comment in state converters.

kuanchihwang committed:
Indicate the originating procedure name in error messages

kuanchihwang commented on schemes/mmm/sf_mynn_compat.F90:
```diff
@@ -0,0 +1,513 @@
+
+        ! This threshold is hardcoded to the same value as in MMM physics.
+        ! It is named `xice_threshold` there.
+        real(kind_phys), parameter :: sea_ice_area_fraction_threshold = 0.02_kind_phys
```
I moved the parameter to module level in https://github.com/ESCOMP/atmospheric_physics/pull/340/commits/3b0af3dcec7181edb22520bee3ca826131ad5e1c.

kuanchihwang committed:
Pin mmm-physics repository by tag instead of hash

kuanchihwang committed:
Centralize option definitions into one place

The definitions of `isfflx`, `isftcflx`, `iz0tlnd`, `spp_pbl`, and `xice_threshold`
have been moved to the `mmm_physics_compat_init` subroutine. This avoids duplicate
definitions in multiple places.

Also adjust misleading code comments.

kuanchihwang committed:
Adjust unit tests for changes in arguments

kuanchihwang committed:
Prefix subroutine name in allocation error messages

kuanchihwang committed:
Prefix subroutine name in allocation error messages retrospectively

kuanchihwang commented on test/unit-test/tests/mmm/mmm_physics_compat_tests.pf:
```diff
@@ -1,3 +1,28 @@
+    ! Should set options that are specific to MMM physics.
+    @assertEqual(1, isfflx)
+    @assertEqual(0, isftcflx)
+    @assertEqual(0, iz0tlnd)
```
As previously mentioned, these constants are defined in the physics driver of standalone MPAS. Therefore, it is not possible to reference them (either through use statements or procedure calls) from here.

nusbaume reviewed (approved):
Thanks for resolving my concerns @kuanchihwang!  I had one last unit update that I missed in the first review, but otherwise everything else looks good to me.

cacraigucar commented on schemes/mmm/mmm_physics_compat.F90:
```diff
@@ -12,22 +13,49 @@ module mmm_physics_compat
 
+        ! This threshold is hardcoded to the same value as in MMM physics.
+        ! It is named `xice_threshold` there.
+        real(kind_phys), parameter :: sea_ice_area_fraction_threshold = 0.02_kind_phys
```
@kuanchihwang - I agree that this can be postponed until a future date when namelists are introduced.

kuanchihwang committed:
The units for variables named "*_area_fraction" should be "fraction"

kuanchihwang commented on schemes/mmm/mmm_physics_compat.meta:
```diff
@@ -23,6 +23,18 @@
+  intent = out
+[ xice_threshold ]
+  standard_name = sea_ice_area_fraction_threshold_for_mmm_scheme
+  units = 1
```
Resolved in https://github.com/ESCOMP/atmospheric_physics/pull/340/commits/0872abb5a8d814684d435d7c6533769c24306811.
